Properly autosave polymorphic has_one associations

piotrj opened the pull request:
Fixes #23266

Current behavior of ActiveRecord is that (unless autosave option
is turned off) it autosaves the has one relation. This means the
following:

```
    class Driver < ActiveRecord::Base
      has_one :car
    end

    class Car < ActiveRecord::Base
      belongs_to :driver
    end

    driver1 = Driver.create!(id: 100)
    driver2 = Driver.create!(id: 143)
    car = Car.create!

    driver1.car = car # at this point car's driver_id is 100
    driver2.car = car # here it changes to 143
    driver1.save! # ActiveRecord autosaves car as well and sets
                  # car.driver_id back to 100

```

The caveat was that only the foreign_key was updated during autosave.
The problem appeared when the belongs_to association was polymorphic.
For example (this can also be seen in the test in the commit):

```
    class Image < ActiveRecord::Base
      belongs_to :imageable, polymorphic: true
    end

    class Post < ActiveRecord::Base
      has_one :image, as: :imageable
    end

    class Person < ActiveRecord::Base
      has_one :image, as: :imageable
    end

    post = Post.create!(id: 155)
    person = Person.create!(id: 204)
    image = Image.create!

    post.image = image # imageable_id set to 155,
                       # imageable_type set to "Post"
    person.image = image # imageable_id set to 204,
                         # imageable_type set to "Person"
    post.save! # only imageable_id is set to 155,
               # but imageable_type is still "Person"
               # result: image.imageable points either to nil
               # or some random Person # if there is Person with id 155

```

This commit fixes this behavior by making sure that if the relation is
polymorphic then except for the foreign_key the foreign_type is also
updated on the record.

piotrj committed:
Properly autosave polymorphic has_one associations

Fixes #23266

Current behavior of ActiveRecord is that (unless autosave option
is turned off) it autosaves the has one relation. This means the
following:

```
    class Driver < ActiveRecord::Base
      has_one :car
    end

    class Car < ActiveRecord::Base
      belongs_to :driver
    end

    driver1 = Driver.create!(id: 100)
    driver2 = Driver.create!(id: 143)
    car = Car.create!

    driver1.car = car # at this point car's driver_id is 100
    driver2.car = car # here it changes to 143
    driver1.save! # ActiveRecord autosaves car as well and sets
                  # car.driver_id back to 100

```

The caveat was that only the foreign_key was updated during autosave.
The problem appeared when the belongs_to association was polymorphic.
For example (this can also be seen in the test in the commit):

```
    class Image < ActiveRecord::Base
      belongs_to :imageable, polymorphic: true
    end

    class Post < ActiveRecord::Base
      has_one :image, as: :imageable
    end

    class Person < ActiveRecord::Base
      has_one :image, as: :imageable
    end

    post = Post.create!(id: 155)
    person = Person.create!(id: 204)
    image = Image.create!

    post.image = image # imageable_id set to 155,
                       # imageable_type set to "Post"
    person.image = image # imageable_id set to 204,
                         # imageable_type set to "Person"
    post.save! # only imageable_id is set to 155,
               # but imageable_type is still "Person"
               # result: image.imageable points either to nil
               # or some random Person # if there is Person with id 155

```

This commit fixes this behavior by making sure that if the relation is
polymorphic then except for the foreign_key the foreign_type is also
updated on the record.

piotrj commented:
@matthewd @maclover7 anything I can do to help with getting this through?

oniofchaos commented:
Hi @piotrj, stumbled upon this through https://www.codetriage.com/. Is this something that still makes sense for Rails? If so, can you fix up the merge conflicts? We can try flagging some other contributors who work with Code Triage to get it merged in.

piotrj commented:
@oniofchaos it seems that this bug still exists. In general the whole idea of autosaving those relations is a little bit weird, but since it's there probably it makes sense that it's more or less consistent. 

Anyway - I rebased the code and fixed it a little bit with new changes that appeared in the codebase since I first introduced the changes.

piotrj commented:
@matthewd any chance of merging that in?
